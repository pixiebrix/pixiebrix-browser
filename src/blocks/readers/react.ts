/*
 * Copyright (C) 2020 Pixie Brix, LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import { createSendScriptMessage } from "@/messaging/chrome";
import { READ_REACT_COMPONENT } from "@/messaging/constants";
import { ReaderOutput } from "@/core";
import { registerFactory } from "@/blocks/readers/factory";

export interface ReactConfig {
  type: "react";
  selector: string;
  traverseUp?: number;
  rootProp?: string;
  waitMillis?: number;
}

export const withReactComponent = createSendScriptMessage<ReaderOutput>(
  READ_REACT_COMPONENT
);

async function doRead(reader: ReactConfig): Promise<ReaderOutput> {
  const { selector, traverseUp = 0, waitMillis = 1000, rootProp } = reader;
  return await withReactComponent({
    selector,
    traverseUp,
    waitMillis,
    rootProp,
  });
}

registerFactory("react", doRead);
