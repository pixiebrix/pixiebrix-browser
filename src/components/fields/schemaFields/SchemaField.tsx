/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useContext } from "react";
import {
  type SchemaFieldProps,
  type SchemaFieldComponent,
} from "@/components/fields/schemaFields/propTypes";
import BasicSchemaField from "@/components/fields/schemaFields/BasicSchemaField";
import AppApiIntegrationDependencyField from "@/components/fields/schemaFields/AppApiIntegrationDependencyField";
import CssClassField from "./CssClassField";
import HeadingStyleField from "./HeadingStyleField";
import {
  isAppServiceField,
  isCssClassField,
  isHeadingStyleField,
  hasCustomWidget,
} from "./fieldTypeCheckers";
import RootAwareField from "@/components/fields/schemaFields/RootAwareField";
import SchemaFieldContext from "@/components/fields/schemaFields/SchemaFieldContext";
import { get } from "lodash";
import defaultFieldFactory from "@/components/fields/schemaFields/defaultFieldFactory";

const SchemaField: SchemaFieldComponent = (props) => {
  const { schema, uiSchema } = props;
  const { customWidgets } = useContext(SchemaFieldContext);

  if (isAppServiceField(schema)) {
    return <AppApiIntegrationDependencyField {...props} />;
  }

  if (isCssClassField(schema)) {
    return <CssClassField {...props} />;
  }

  if (isHeadingStyleField(schema)) {
    return <HeadingStyleField {...props} />;
  }

  if (hasCustomWidget(uiSchema)) {
    const widget = get(customWidgets, uiSchema["ui:widget"] as string);

    // If the uiWidget is registered, render it. Otherwise, render the default field.
    if (widget) {
      const Component = defaultFieldFactory(
        widget as React.VFC<SchemaFieldProps>
      );
      return <Component {...props} />;
    }
  }

  if (props.name.endsWith(".isRootAware")) {
    // Hide the isRootAware field if rendered as part of config.isRootAware. The field was introduced for
    // backward compatibility when upgrading DOM bricks to be root-aware.
    return <RootAwareField {...props} />;
  }

  return <BasicSchemaField {...props} />;
};

export default SchemaField;
