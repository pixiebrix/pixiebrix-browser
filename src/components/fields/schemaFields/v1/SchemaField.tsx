/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { SchemaFieldComponent } from "@/components/fields/schemaFields/propTypes";
import React, { useContext, useMemo } from "react";
import SchemaFieldContext, {
  getDefaultField,
} from "@/components/fields/schemaFields/SchemaFieldContext";

/**
 * A schema-based field that automatically determines it's layout/widget based on the schema and uiSchema.
 *
 * @see SchemaFieldContext
 * @see getDefaultField
 */
const SchemaField: SchemaFieldComponent = (props) => {
  const { schema } = props;
  const { customFields } = useContext(SchemaFieldContext);

  const Field = useMemo(() => {
    const overrideField = customFields.find((x) => x.match(schema))?.Component;
    return overrideField ?? getDefaultField(schema);
  }, [customFields, schema]);

  return <Field {...props} />;
};

export default SchemaField;
