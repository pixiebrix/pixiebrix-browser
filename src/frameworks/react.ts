/*
 * Copyright (C) 2020 Pixie Brix, LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import fromPairs from "lodash/fromPairs";

interface DOMFiber {
  _instance: any;
  _currentElement: {
    _owner: DOMFiber;
  };
}

interface ComponentFiber {
  return: ComponentFiber;
  memoizedProps: { [key: string]: unknown };
}

export class ComponentNotFoundError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ComponentNotFoundError";
  }
}

export function readReactProps(
  fiber: ComponentFiber,
  rootProp?: string
): Record<string, unknown> {
  // getComponentFiber is already returning the "return" for the component
  const memoizedProps = fiber.memoizedProps;
  console.debug("Memoized props", memoizedProps);
  const props = rootProp ? memoizedProps[rootProp] : memoizedProps;
  return fromPairs(Object.entries(props).filter(([key]) => key !== "children"));
}

export function findReactComponent(
  dom: HTMLElement,
  traverseUp = 0
): ComponentFiber {
  // https://stackoverflow.com/questions/29321742/react-getting-a-component-from-a-dom-element-for-debugging
  // https://stackoverflow.com/a/39165137/402560

  const key = Object.keys(dom).find((key) =>
    key.startsWith("__reactInternalInstance$")
  );
  // @ts-ignore: works in practice
  const domFiber: DOMFiber | null = dom[key];

  if (domFiber == null) {
    throw new ComponentNotFoundError("React fiber not found at selector");
  }

  // react <16
  // TODO: find a react 15.0 app to test this on
  if (domFiber._currentElement) {
    throw new Error("Support for React <16 not implemented");
    // let componentFiber = domFiber._currentElement._owner;
    // for (let i = 0; i < traverseUp; i++) {
    //     componentFiber = componentFiber._currentElement._owner;
    // }
    // return componentFiber._instance;
  }

  // react 16+
  const getComponentFiber = (fiber: any) => {
    //return fiber._debugOwner; // this also works, but is __DEV__ only
    let parentFiber = fiber.return;
    while (typeof parentFiber.type == "string") {
      parentFiber = parentFiber.return;
    }
    return parentFiber;
  };
  let componentFiber = getComponentFiber(domFiber);
  for (let i = 0; i < traverseUp; i++) {
    componentFiber = getComponentFiber(componentFiber);
  }
  return componentFiber;
}
