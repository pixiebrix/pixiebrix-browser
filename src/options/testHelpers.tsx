/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore } from "@reduxjs/toolkit";
import { authSlice } from "@/auth/authSlice";
import extensionsSlice from "@/store/extensionsSlice";
import settingsSlice from "@/store/settingsSlice";
import { blueprintModalsSlice } from "@/options/pages/blueprints/modals/blueprintModalsSlice";
import { createRenderWithWrappers } from "@/testUtils/testHelpers";
import blueprintsSlice from "@/options/pages/blueprints/blueprintsSlice";
import { recipesSlice } from "@/recipes/recipesSlice";

const renderWithWrappers = createRenderWithWrappers(() =>
  configureStore({
    reducer: {
      auth: authSlice.reducer,
      settings: settingsSlice.reducer,
      options: extensionsSlice.reducer,
      blueprintModals: blueprintModalsSlice.reducer,
      blueprints: blueprintsSlice.reducer,
      recipes: recipesSlice.reducer,
    },
  })
);

// eslint-disable-next-line import/export -- re-export RTL
export * from "@testing-library/react";
// eslint-disable-next-line import/export -- override render
export { renderWithWrappers as render };
