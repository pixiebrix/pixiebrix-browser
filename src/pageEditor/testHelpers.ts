/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore } from "@reduxjs/toolkit";
import { authSlice } from "@/auth/authSlice";
import extensionsSlice from "@/store/extensionsSlice";
import servicesSlice from "@/store/servicesSlice";
import settingsSlice from "@/store/settingsSlice";
import { editorSlice } from "@/pageEditor/slices/editorSlice";
import sessionSlice from "@/pageEditor/slices/sessionSlice";
import { savingExtensionSlice } from "@/pageEditor/panes/save/savingExtensionSlice";
import runtimeSlice from "@/pageEditor/slices/runtimeSlice";
import { logSlice } from "@/components/logViewer/logSlice";
import {
  createRenderHookWithWrappers,
  createRenderWithWrappers,
} from "@/testUtils/testHelpers";
import analysisSlice from "@/analysis/analysisSlice";
import pageEditorAnalysisManager from "./analysisManager";
import { tabStateSlice } from "@/pageEditor/tabState/tabStateSlice";
import { appApi } from "@/services/api";
import { modDefinitionsSlice } from "@/modDefinitions/modDefinitionsSlice";
import { modDefinitionsMiddleware } from "@/modDefinitions/modDefinitionsListenerMiddleware";

const configureStoreForTests = () =>
  configureStore({
    reducer: {
      auth: authSlice.reducer,
      options: extensionsSlice.reducer,
      services: servicesSlice.reducer,
      settings: settingsSlice.reducer,
      editor: editorSlice.reducer,
      session: sessionSlice.reducer,
      savingExtension: savingExtensionSlice.reducer,
      runtime: runtimeSlice.reducer,
      logs: logSlice.reducer,
      analysis: analysisSlice.reducer,
      tabState: tabStateSlice.reducer,
      modDefinitions: modDefinitionsSlice.reducer,
      [appApi.reducerPath]: appApi.reducer,
    },
    middleware(getDefaultMiddleware) {
      /* eslint-disable unicorn/prefer-spread -- It's not Array#concat, can't use spread */
      return getDefaultMiddleware()
        .concat(appApi.middleware)
        .concat(pageEditorAnalysisManager.middleware)
        .concat(modDefinitionsMiddleware);
      /* eslint-enable unicorn/prefer-spread */
    },
  });

const renderWithWrappers = createRenderWithWrappers(configureStoreForTests);
const renderHookWithWrappers = createRenderHookWithWrappers(
  configureStoreForTests
);

// eslint-disable-next-line import/export -- re-export RTL
export * from "@testing-library/react";
// eslint-disable-next-line import/export -- override render
export { renderWithWrappers as render };
export { renderHookWithWrappers as renderHook };
