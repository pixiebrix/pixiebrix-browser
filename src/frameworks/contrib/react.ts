/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { pickBy } from "lodash";
import { ComponentNotFoundError, ignoreNotFound } from "@/frameworks/errors";
import { RootInstanceVisitor } from "@/frameworks/scanner";
import { ReadableComponentAdapter, traverse } from "@/frameworks/component";
import { isNode } from "@/frameworks/dom";

// React architecture references:
// https://github.com/acdlite/react-fiber-architecture
// https://indepth.dev/posts/1008/inside-fiber-in-depth-overview-of-the-new-reconciliation-algorithm-in-react#memoizedprops
// https://github.com/Venryx/mobx-devtools-advanced/blob/master/Docs/TreeTraversal.md
// https://stackoverflow.com/questions/29321742/react-getting-a-component-from-a-dom-element-for-debugging

interface RootInstance {
  _reactRootContainer: unknown;
}

/**
 * Structure from:
 * - https://github.com/facebook/react/blob/50d9451f320a9aaf94304209193562cc385567d8/packages/react-devtools-shared/src/backend/legacy/renderer.js#L739
 * - https://github.com/facebook/react/blob/v15.0.0/src/renderers/shared/reconciler/ReactReconciler.js
 */
interface LegacyInstance {
  /**
   * Public instance
   */
  _instance: unknown;

  _currentElement: {
    _owner: LegacyInstance;
  };
}

/**
 * Type definition from: https://github.com/facebook/react/blob/4c6470cb3b821f3664955290cd4c4c7ac0de733a/packages/react-reconciler/src/ReactInternalTypes.js#L62
 */
interface Fiber {
  /**
   * The Fiber to return to after finishing processing this one.
   * This is effectively the parent, but there can be multiple parents (two)
   * so this is only the parent of the thing we're currently processing.
   * It is conceptually the same as the return address of a stack frame.
   */
  return: Fiber | null;

  /**
   * The props used to create the output.
   */
  memoizedProps: { [prop: string]: unknown };

  /**
   * The resolved function/class associated with this fiber.
   */
  type: string | object;

  /**
   * The local state associated with this fiber.
   */
  stateNode: Node | object;
}

export function isManaged(node: Node): boolean {
  return Object.keys(node).some((key) =>
    key.startsWith("__reactInternalInstance$")
  );
}

export function hasReactProps(fiber: Fiber): boolean {
  return Object.keys(fiber.memoizedProps).some((x) => x !== "children");
}

export function readReactProps(fiber: Fiber): { [prop: string]: unknown } {
  return pickBy(fiber.memoizedProps, (value, key) => key !== "children");
}

function getComponentFiber(fiber: Fiber): Fiber {
  // return fiber._debugOwner; // this also works, but is __DEV__ only
  let parentFiber = fiber.return;
  while (typeof parentFiber.type === "string") {
    // string for HTML nodes, so traverse
    parentFiber = parentFiber.return;
  }
  return parentFiber;
}

export function findReactComponent(node: Node, traverseUp = 0): Fiber {
  // https://stackoverflow.com/a/39165137/402560
  const key = Object.keys(node).find((key) =>
    key.startsWith("__reactInternalInstance$")
  );

  const domFiber: Fiber | LegacyInstance | null = (node as any)[key];

  if (domFiber == null) {
    throw new ComponentNotFoundError("React fiber not found for element");
  }

  if ("_currentElement" in domFiber) {
    // FIXME: test this for React <16
    const owner = (x: LegacyInstance) => x._currentElement._owner;
    const fiber = traverse(owner, owner(domFiber), traverseUp);
    return fiber._instance as Fiber;
  } else {
    return traverse(getComponentFiber, getComponentFiber(domFiber), traverseUp);
  }
}

export class ReactRootVisitor implements RootInstanceVisitor<RootInstance> {
  public rootInstances: RootInstance[] = [];
  visit(node: Element | Node): boolean {
    if ("_reactRootContainer" in node) {
      this.rootInstances.push(node);
      return false;
    }
    return true;
  }
}

export const adapter: ReadableComponentAdapter<Fiber> = {
  isManaged,
  getComponent: (node) => ignoreNotFound(() => findReactComponent(node, 0)),
  getParent: getComponentFiber,
  getNode: (instance) =>
    isNode(instance.stateNode) ? instance.stateNode : null,
  getData: readReactProps,
  hasData: hasReactProps,
};

export default adapter;
