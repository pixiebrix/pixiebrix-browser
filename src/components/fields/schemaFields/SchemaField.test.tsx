/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import SchemaField from "@/components/fields/schemaFields/SchemaField";
import { Formik } from "formik";
import { ApiVersion, Expression, Schema, TemplateEngine } from "@/core";
import { createFormikTemplate } from "@/tests/formHelpers";
import { waitForEffect } from "@/tests/testHelpers";
import userEvent from "@testing-library/user-event";
import { uniq } from "lodash";

async function expectToggleOptions(container: HTMLElement, expected: string[]) {
  // React Bootstrap dropdown does not render children items unless toggled
  userEvent.click(container.querySelector("button"));
  const actual = new Set(
    [...container.querySelectorAll("a")].map((x) =>
      x.getAttribute("data-testid")
    )
  );
  await waitFor(() => {
    expect(actual).toEqual(new Set(expected));
  });
}

interface SchemaTestCase {
  name: string;
  schema: Schema;
}

const sampleSchemas: SchemaTestCase[] = [
  {
    name: "empty",
    schema: {},
  },

  {
    name: "any properties",
    schema: {
      additionalProperties: true,
    },
  },

  {
    name: "basic string",
    schema: {
      type: "string",
    },
  },
  {
    name: "basic number",
    schema: {
      type: "number",
    },
  },
  {
    name: "basic integer",
    schema: {
      type: "integer",
    },
  },
  {
    name: "basic boolean",
    schema: {
      type: "boolean",
    },
  },

  {
    name: "object with defined properties",
    schema: {
      type: "object",
      properties: {
        myString: { type: "string" },
        myBool: { type: "boolean" },
      },
    },
  },
  {
    name: "object with any additional properties",
    schema: {
      type: "object",
      additionalProperties: true,
    },
  },
  {
    name: "object with property types",
    schema: {
      type: "object",
      additionalProperties: {
        type: ["string", "number", "boolean"],
      },
    },
  },
  {
    name: "object with required fields",
    schema: {
      type: "object",
      properties: {
        myRequiredString: { type: "string" },
        myRequiredBool: { type: "boolean" },
        myString: { type: "string" },
        myNumber: { type: "number" },
      },
      required: ["myRequiredString", "myRequiredBool"],
    },
  },

  {
    name: "string array",
    schema: {
      type: "array",
      items: { type: "string" },
    },
  },
  {
    name: "array of objects with properties",
    schema: {
      type: "array",
      items: {
        type: "object",
        properties: {
          name: { type: "string" },
          value: { type: ["string", "number", "boolean"] },
        },
      },
    },
  },

  {
    name: "oneOf string or number",
    schema: {
      oneOf: [{ type: "string" }, { type: "number" }],
    },
  },
  {
    name: "oneOf boolean or number",
    schema: {
      oneOf: [{ type: "boolean" }, { type: "number" }],
    },
  },
  {
    name: "oneOf boolean or string",
    schema: {
      oneOf: [{ type: "boolean" }, { type: "string" }],
    },
  },
  {
    name: "oneOf boolean, string, or number",
    schema: {
      oneOf: [{ type: "boolean" }, { type: "string" }, { type: "number" }],
    },
  },
];

const schemaTestCases: ReadonlyArray<
  [name: string, schema: Schema]
> = sampleSchemas.map(({ name, schema }) => [name, schema]);

function expressionValue<T extends TemplateEngine>(
  type: T,
  value = ""
): Expression<string, T> {
  return {
    __type__: type,
    __value__: value,
  };
}

describe("SchemaField", () => {
  test.each([["v1"], ["v2"]])(
    "don't show toggle widget for %s",
    (version: ApiVersion) => {
      const { container } = render(
        <Formik
          onSubmit={() => {}}
          initialValues={{ apiVersion: version, testField: "" }}
        >
          <SchemaField
            name="testField"
            schema={{
              type: "string",
              title: "Test Field",
              description: "A test field",
            }}
          />
        </Formik>
      );

      // Renders text entry HTML element
      expect(container.querySelector("textarea")).not.toBeNull();
      expect(container.querySelector("button")).toBeNull();
    }
  );

  test("string field options", async () => {
    const { container } = render(
      <Formik
        onSubmit={() => {}}
        initialValues={{ apiVersion: "v3", testField: "" }}
      >
        <SchemaField
          name="testField"
          schema={{
            type: "string",
            title: "Test Field",
            description: "A test field",
          }}
        />
      </Formik>
    );

    // Renders text entry HTML element
    expect(container.querySelector("textarea")).not.toBeNull();

    await expectToggleOptions(container, ["string", "var", "omit"]);
  });

  test("integer field options", async () => {
    const { container } = render(
      <Formik
        onSubmit={() => {}}
        initialValues={{ apiVersion: "v3", testField: 42 }}
      >
        <SchemaField
          name="testField"
          schema={{
            type: "integer",
            title: "Test Field",
            description: "A test field",
          }}
        />
      </Formik>
    );

    // Renders number entry HTML element
    expect(container.querySelector("input[type='number']")).not.toBeNull();
    await expectToggleOptions(container, ["number", "var", "omit"]);
  });

  test.each`
    startValue                            | inputMode     | toggleOption  | expectedEndValue
    ${{ foo: "bar" }}                     | ${"Object"}   | ${"Variable"} | ${expressionValue("var")}
    ${expressionValue("var", "abc")}      | ${"Variable"} | ${"Object"}   | ${{}}
    ${expressionValue("var", "abc")}      | ${"Variable"} | ${"Text"}     | ${expressionValue("nunjucks")}
    ${expressionValue("nunjucks", "def")} | ${"Text"}     | ${"Array"}    | ${[]}
  `(
    "Test field toggle transition from $inputMode to $toggleOption",
    async ({ startValue, toggleOption, expectedEndValue }) => {
      const initialState = {
        apiVersion: "v3",
        myField: startValue,
      };
      const onSubmit = jest.fn();
      const FormikTemplate = createFormikTemplate(initialState, onSubmit);

      // Using an empty schema to allow anything, since we're testing toggling, not schema parsing
      render(
        <FormikTemplate>
          <SchemaField name={"myField"} schema={{}} />
        </FormikTemplate>
      );

      await waitForEffect();

      const toggle = screen
        .getByTestId("toggle-myField")
        .querySelector("button");
      expect(toggle).not.toBeNull();

      userEvent.click(toggle);

      const newOption = screen.getByText(toggleOption, { exact: false });
      expect(newOption).not.toBeNull();

      // Await this element to avoid the "unable to click element" error
      await waitFor(() => {
        userEvent.click(newOption);
      });

      userEvent.click(screen.getByRole("button", { name: /submit/i }));

      await waitFor(() => {
        expect(onSubmit).toHaveBeenCalledWith({
          apiVersion: "v3",
          myField: expectedEndValue,
        });
      });
    }
  );

  test("string/integer field options", async () => {
    const { container } = render(
      <Formik
        onSubmit={() => {}}
        initialValues={{ apiVersion: "v3", testField: 42 }}
      >
        <SchemaField
          name="testField"
          schema={{
            type: ["integer", "string"],
            title: "Test Field",
            description: "A test field",
          }}
        />
      </Formik>
    );

    // Renders number entry HTML element because current value is a number
    expect(container.querySelector("input[type='number']")).not.toBeNull();
    await expectToggleOptions(container, ["string", "number", "var", "omit"]);
  });

  test("v2 field oneOf type priority shows text", () => {
    const FormikTemplate = createFormikTemplate({ apiVersion: "v2" });
    const schema: Schema = {
      oneOf: [{ type: "boolean" }, { type: "string" }, { type: "number" }],
    };
    const { container } = render(
      <FormikTemplate>
        <SchemaField name="myField" schema={schema} />
      </FormikTemplate>
    );

    // Should render textarea for anything that includes text
    expect(container.querySelector("textarea")).not.toBeNull();
  });

  test("v2 field type array shows text", () => {
    const FormikTemplate = createFormikTemplate({ apiVersion: "v2" });
    const schema: Schema = {
      type: ["boolean", "number", "string"],
    };
    const { container } = render(
      <FormikTemplate>
        <SchemaField name="myField" schema={schema} />
      </FormikTemplate>
    );

    // Should render textarea for anything that includes text
    expect(container.querySelector("textarea")).not.toBeNull();
  });

  test.each(schemaTestCases)(
    "v3 field toggle doesn't show duplicate options - %s",
    async (_, schema) => {
      const FormikTemplate = createFormikTemplate({ apiVersion: "v3" });
      const { container } = render(
        <FormikTemplate>
          <SchemaField name="aTestField" schema={schema} />
        </FormikTemplate>
      );

      await waitForEffect();

      const toggle = screen
        .getByTestId("toggle-aTestField")
        .querySelector("button");
      expect(toggle).not.toBeNull();

      userEvent.click(toggle);

      await waitFor(() => {
        const testIds = [
          ...container.querySelectorAll("a.dropdown-item"),
        ].map((x) => x.getAttribute("data-testid"));
        expect(testIds).toEqual(uniq(testIds));
      });
    }
  );

  test.each(schemaTestCases)(
    "v3 field toggle always renders 'omit' last - %s",
    async (_, schema) => {
      const FormikTemplate = createFormikTemplate({ apiVersion: "v3" });
      const { container } = render(
        <FormikTemplate>
          <SchemaField name="aTestField" schema={schema} />
        </FormikTemplate>
      );

      await waitForEffect();

      const toggle = screen
        .getByTestId("toggle-aTestField")
        .querySelector("button");
      expect(toggle).not.toBeNull();

      userEvent.click(toggle);

      await waitFor(() => {
        const testIds = [
          ...container.querySelectorAll("a.dropdown-item"),
        ].map((x) => x.getAttribute("data-testid"));
        if (testIds.includes("omit")) {
          expect(testIds[testIds.length - 1]).toEqual("omit");
        }
      });
    }
  );
});
