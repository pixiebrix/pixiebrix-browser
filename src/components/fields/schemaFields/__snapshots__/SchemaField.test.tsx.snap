// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchemaField labelled enum field schema defaults to selection widget 1`] = `
<DocumentFragment>
  <div
    class="formGroup form-group row"
  >
    <label
      class="label form-label col-form-label col-xl-2 col-lg-3"
      for="testField"
    >
      Test Field
    </label>
    <div
      class="col-xl-10 col-lg-9"
    >
      <div
        class="root"
      >
        <div
          class="field"
        >
          <textarea
            class="form-control"
            id="testField"
            name="testField"
            rows="1"
          >
            foo
          </textarea>
        </div>
        <div
          class="dropdown show dropdown"
          data-test-selected="Text"
          data-testid="toggle-testField"
        >
          <button
            aria-expanded="true"
            aria-haspopup="true"
            class="dropdown-toggle btn btn-link"
            type="button"
          >
            <span
              class="symbol"
            >
              <test-file-stub
                classname="root"
              />
            </span>
          </button>
          <div
            aria-labelledby=""
            class="dropdown-menu show dropdown-menu-right"
            data-popper-escaped="true"
            data-popper-placement="bottom-end"
            data-popper-reference-hidden="true"
            style="position: absolute; top: 0px; left: 0px; margin: 0px; transform: translate(0px, 0px);"
            x-placement="bottom-end"
          >
            <a
              class="dropdown-item"
              data-testid="select"
              href="#"
              role="button"
            >
              <span
                class="symbol"
              >
                <test-file-stub
                  classname="root"
                />
              </span>
               Select...
            </a>
            <a
              class="dropdown-item"
              data-testid="string"
              href="#"
              role="button"
            >
              <span
                class="symbol"
              >
                <test-file-stub
                  classname="root"
                />
              </span>
               Text
            </a>
            <a
              class="dropdown-item"
              data-testid="var"
              href="#"
              role="button"
            >
              <span
                class="symbol"
              >
                <test-file-stub
                  classname="root"
                />
              </span>
               Variable
            </a>
          </div>
        </div>
      </div>
      <small
        class="text-muted form-text"
      >
        <span>
          A test field
        </span>
      </small>
    </div>
  </div>
</DocumentFragment>
`;
